dspy/clients/lm.py

```diff
@@ -151,3 +229,114 @@ def _inspect_history(lm, n: int = 1):
             print(_red(choices_text, end=""))
 
     print("\n\n\n")
+
+# TODO: This part can be moved to a separate module
+
+from dspy.clients.openai import (
+    FinetuneJobOpenAI,
+    finetune_openai,
+)
+from dspy.clients.anyscale import (
+    FinetuneJobAnyScale,
+    finetune_anyscale,
+)
+
+
+_PROVIDER_ANYSCALE = "anyscale"
+_PROVIDER_OPENAI = "openai"
+
+
+def get_provider_finetune_job_class(provider: str) -> Type[FinetuneJob]:
+    """Get the FinetuneJob class for the provider."""
+    # Mapping from provider to finetune job type
+    _PROVIDER_TO_FINETUNE_JOB_CLASS = {
+        _PROVIDER_ANYSCALE: FinetuneJobAnyScale,
+        _PROVIDER_OPENAI: FinetuneJobOpenAI,
+    }
+
+    # Get the FinetuneJob class for the provider
+    _CLS = _PROVIDER_TO_FINETUNE_JOB_CLASS[provider]
+
+    return _CLS
+
+
+def get_provider_finetune_function(provider: str) -> callable:
+    """Return the finetune function for the given model."""
+    # Mapping from provider to finetune function
+    _PROVIDER_TO_FINETUNE_FUNCTION = {
+        _PROVIDER_ANYSCALE: finetune_anyscale,
+        _PROVIDER_OPENAI: finetune_openai,
+    }
+
+    # Get the finetuning provider
+    finetune_function = _PROVIDER_TO_FINETUNE_FUNCTION[provider]
+
+    return finetune_function
+
+
+def execute_finetune_job(
+    job: FinetuneJob[Type[LM]],
+    lm: LM,
+    cache_finetune: bool=True
+):
+    """Execute the finetune job in a blocking manner."""
+    try:
+        job_kwargs = job.get_kwargs()
+        if cache_finetune:
+            model = cached_finetune(job=job, **job_kwargs)
+        else:
+            model = finetune(job=job, **job_kwargs)
+        lm = lm.copy(model=model)
+        job.set_result(lm)
+    except Exception as err:
+        logger.error(err)
+        job.set_result(err)
+
+
+# TODO: Perhaps we shouldn't directly cache the data
+# TODO: Add DiskCache, ignore job
+
+def cached_finetune(
```

Comment:
curious - what does this `cached_finetune ` do? from the code it's just a tunnel to `finetune` method

Reply:
Oh, yes, this is meant to be the cached version of a function that actually does the work. Similar to the relationship between the following two functions: 

```
@functools.lru_cache(maxsize=None)
def cached_litellm_completion(request):
    return litellm_completion(request, cache={"no-cache": False, "no-store": False})


def litellm_completion(request, cache={"no-cache": True, "no-store": True}):
    kwargs = ujson.loads(request)
    return litellm.completion(cache=cache, **kwargs)
```